File: deno-knowledge-app/src/cache.ts
// File-backed aggregate cache with periodic background refresh.
// Caches the EXPENSIVE shared data: the global layer (skills/agents/hooks/global
// CLAUDE.md — identical for every project) + the multi-project sidebar (open
// tasks + 7d cost per project). The currently-selected project DETAIL stays live
// (computed per request in server.ts), never cached.

import { dirname, join } from "@std/path";
import { collectGlobal } from "./collectors/index.ts";
import { collectSidebar, type SidebarProject } from "./collectors/sidebar.ts";
import { buildContext } from "./context.ts";

const HOME = Deno.env.get("HOME") ?? "/tmp";

// NOTE: tn (TaskNotes) is intentionally NOT aggregated here. The vault tn system
// spans customer-specific (kunde) and personal projects; aggregating it would
// process exactly the data the org rules forbid. The cross-project view stays
// limited to repo backlog (dev tasks) + token cost.
export interface Aggregate {
  generated_at: number; // epoch ms
  global: Record<string, unknown>;
  projects: SidebarProject[];
}

function cacheFile(): string {
  return Deno.env.get("CC_KNOWLEDGE_CACHE") ??
    join(HOME, ".cache", "cc-knowledge", "cache.json");
}

function ttlMs(): number {
  const v = Number(Deno.env.get("CC_KNOWLEDGE_TTL_MS"));
  return Number.isFinite(v) && v > 0 ? v : 15 * 60 * 1000;
}

/** Pure freshness predicate (testable without IO). */
export function isFresh(agg: Aggregate | null, ttl: number, now: number): boolean {
  return !!agg && (now - agg.generated_at) < ttl;
}

/** Pure freshness predicate fuer Per-Projekt-Cache-Eintraege. */
export function isProjectContextFresh(
  entry: { generated_at: number; context: Record<string, unknown> } | undefined,
  ttl: number,
  now: number,
): boolean {
  return !!entry && (now - entry.generated_at) < ttl;
}

export async function readCacheFile(path = cacheFile()): Promise<Aggregate | null> {
  try {
    const obj = JSON.parse(await Deno.readTextFile(path)) as Aggregate;
    if (obj && typeof obj.generated_at === "number" && Array.isArray(obj.projects)) return obj;
    return null;
  } catch {
    return null;
  }
}

export async function writeCacheFile(agg: Aggregate, path = cacheFile()): Promise<void> {
  try {
    await Deno.mkdir(dirname(path), { recursive: true });
    await Deno.writeTextFile(path, JSON.stringify(agg));
  } catch { /* best-effort; a missing cache just means recompute next refresh */ }
}

async function computeAggregate(
  activeRepo: string,
  home: string,
  now: number,
): Promise<Aggregate> {
  return {
    generated_at: now,
    global: await collectGlobal(home),
    projects: await collectSidebar(activeRepo),
  };
}

let current: Aggregate | null = null;
let refreshing: Promise<void> | null = null;
let started = false; // true once a real server primed the cache (not in tests)

/** Latest aggregate (in-memory). */
export function getAggregate(): Aggregate | null {
  return current;
}

/** True while an aggregate refresh (the cross-project scan) is in flight. */
function isRefreshing(): boolean {
  return refreshing !== null;
}

/**
 * Single-flight refresh: at most ONE aggregate computation runs at a time across
 * the whole process. Concurrent callers (boot prime + per-request lazy trigger)
 * all await the SAME in-flight run instead of starting their own — this prevents
 * overlapping cross-project scans (the original cause of the RAM blow-up, when
 * overlapping batches each spawned session_analyze.py). The sidebar now reads
 * session JSONLs natively (sessions_native.ts), so this path spawns no Python.
 */
function refreshAggregate(activeRepo: string, home: string): Promise<void> {
  if (refreshing) return refreshing;
  refreshing = (async () => {
    try {
      const next = await computeAggregate(activeRepo, home, Date.now());
      current = next;
      await writeCacheFile(next);
    } catch { /* keep last good aggregate */ }
  })().finally(() => {
    refreshing = null;
  });
  return refreshing;
}

/**
 * Lazy trigger for request handlers: kick off a refresh ONLY if the cache is
 * stale/missing AND none is already running. Non-blocking, fire-and-forget.
 */
export function ensureFresh(activeRepo: string, home: string): void {
  if (!started) return; // never trigger heavy scans from a bare createHandler (tests)
  if (refreshing) return;
  if (isFresh(current, ttlMs(), Date.now())) return;
  refreshAggregate(activeRepo, home);
}

/**
 * Prime once on boot: load a fresh cache file if present (no Python spawned),
 * otherwise compute one aggregate (single-flight). NO background timer — staleness
 * is handled lazily via ensureFresh() on request, so nothing churns while idle.
 */
async function startCache(activeRepo: string, home: string): Promise<void> {
  started = true;
  const fromFile = await readCacheFile();
  if (isFresh(fromFile, ttlMs(), Date.now())) {
    current = fromFile;
    return;
  }
  await refreshAggregate(activeRepo, home);
}

// ---------------------------------------------------------------------------
// Per-Projekt-Detail-Cache (B2)
// ---------------------------------------------------------------------------
// Jedes Projekt bekommt einen eigenen Cache-Eintrag (TTL + single-flight).
// Verhindert, dass bei jedem GET /cc-setup der teure buildContext (inkl.
// collectTokens mit JSONL-Parsing) komplett neu durchlaeuft.

interface ProjectContextEntry {
  generated_at: number; // epoch ms
  context: Record<string, unknown>;
}

// In-memory Projekt-Cache (kein File-Backing noetig — fluechtiger Prozess-Cache)
const projectContextCache = new Map<string, ProjectContextEntry>();
// Single-flight: aktive Berechnungen pro Projekt
const projectContextInFlight = new Map<string, Promise<Record<string, unknown>>>();
// Generations-Zaehler pro Projekt: wird bei jeder Invalidierung inkrementiert.
// Ein in-flight computeFn merkt sich die Generation bei Start und schreibt das
// Ergebnis NUR in den Cache wenn die Generation unveraendert ist — sonst Verwerfen.
const projectContextGeneration = new Map<string, number>();

/**
 * Liefert den gecachten Projekt-Kontext oder startet eine single-flight Berechnung.
 * `projectKey` = aufgeloester Projektpfad (absoluter Pfad) fuer Konsistenz.
 * `computeFn` = die tatsaechliche buildContext-Berechnung.
 */
export async function getProjectContext(
  projectKey: string,
  computeFn: () => Promise<Record<string, unknown>>,
): Promise<Record<string, unknown>> {
  if (!started) {
    // Im Test-Modus (started=false) immer live berechnen — kein Cache
    return await computeFn();
  }

  const now = Date.now();
  const cached = projectContextCache.get(projectKey);
  if (isProjectContextFresh(cached, ttlMs(), now)) {
    return cached!.context;
  }

  // Single-flight: wenn bereits eine Berechnung laeuft, dieselbe awaiten
  const existing = projectContextInFlight.get(projectKey);
  if (existing) return existing;

  // Snapshot der aktuellen Generation: wird nach Abschluss mit dem Ist-Stand
  // verglichen. Hat invalidateProjectContext inzwischen inkrementiert, ist das
  // Ergebnis veraltet → verwerfen, nicht cachen (Race-Fix).
  const genAtStart = projectContextGeneration.get(projectKey) ?? 0;

  const promise = (async () => {
    try {
      const ctx = await computeFn();
      // Nur schreiben wenn keine Invalidierung waehrend des Laufs kam
      if ((projectContextGeneration.get(projectKey) ?? 0) === genAtStart) {
        projectContextCache.set(projectKey, { generated_at: Date.now(), context: ctx });
      }
      return ctx;
    } finally {
      projectContextInFlight.delete(projectKey);
    }
  })();
  projectContextInFlight.set(projectKey, promise);
  return promise;
}

/**
 * Invalidiert den Per-Projekt-Cache-Eintrag — nach Mutations-Aktionen
 * (commit/push/merge/delete/task-status) aufrufen, damit die naechste
 * Anfrage frische Daten sieht statt des alten Caches.
 */
export function invalidateProjectContext(projectKey: string): void {
  projectContextCache.delete(projectKey);
  // Generations-Zaehler inkrementieren: eine in-flight computeFn liest den
  // Zaehler nach Abschluss erneut und verwirft ihr Ergebnis, wenn es sich
  // veraendert hat — so schreibt ein "zu spaet" abgeschlossener Lauf keine
  // veralteten Daten in den Cache (Race-Fix).
  projectContextGeneration.set(
    projectKey,
    (projectContextGeneration.get(projectKey) ?? 0) + 1,
  );
}

/**
 * Boot-Prime fuer das Start-Projekt: einmalig nach startCache aufrufen.
 * Berechnet den Projekt-Kontext fuer das `--cwd`-Repo, damit die erste
 * GET-Anfrage sofort aus dem Cache bedient werden kann.
 * KEIN Prime aller Projekte (vermeide N×-schwere Laeufe).
 */
async function primeProjectContext(
  cwd: string,
  claudeHome: string,
  agg: Aggregate | null,
): Promise<void> {
  if (!started) return;
  // Nur priemen wenn nicht bereits frisch
  if (isProjectContextFresh(projectContextCache.get(cwd), ttlMs(), Date.now())) return;

  try {
    await getProjectContext(cwd, () =>
      buildContext(cwd, claudeHome, {
        projects: agg?.projects ?? [],
        active_project: cwd.split("/").pop()!,
        global: agg?.global,
        skipProject: false,
      }));
  } catch {
    // Boot-Prime ist best-effort; Fehler werden still ignoriert
  }
}

// Fuer Tests: Cache-State resetten
export function _resetProjectContextCacheForTest(): void {
  projectContextCache.clear();
  projectContextInFlight.clear();
  projectContextGeneration.clear();
}

/** Nur fuer Tests: `started`-Flag setzen um Caching-Pfad zu aktivieren. */
export function _setStartedForTest(value: boolean): void {
  started = value;
}

File: go-knowledge-app/internal/cache/cache.go
// Package cache is the file-backed aggregate cache + per-project context cache —
// Go-native port of cache.ts. The Deno single-flight (shared promise) becomes a
// mutex + done-channel; otherwise behavior (TTL, lazy refresh, generation-guarded
// writes) matches.
//
// NOTE: tn (TaskNotes) is intentionally NOT aggregated here (Org rule) — the
// cross-project view stays limited to repo backlog + token cost.
package cache

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"sync"
	"time"

	"go-knowledge-app/internal/appctx"
	"go-knowledge-app/internal/collectors"
)

// Aggregate caches the expensive shared data: the global layer + the
// multi-project sidebar.
type Aggregate struct {
	GeneratedAt int64                       `json:"generated_at"` // epoch ms
	Global      map[string]any              `json:"global"`
	Projects    []collectors.SidebarProject `json:"projects"`
}

func nowMs() int64 { return time.Now().UnixMilli() }

// CacheFile returns the on-disk cache path (CC_KNOWLEDGE_CACHE overrides).
func CacheFile() string {
	if env := os.Getenv("CC_KNOWLEDGE_CACHE"); env != "" {
		return env
	}
	return filepath.Join(collectors.Home(), ".cache", "cc-knowledge", "cache.json")
}

// TTLMs returns the cache TTL (CC_KNOWLEDGE_TTL_MS overrides; default 15 min).
func TTLMs() int64 {
	if env := os.Getenv("CC_KNOWLEDGE_TTL_MS"); env != "" {
		if v, err := strconv.ParseInt(env, 10, 64); err == nil && v > 0 {
			return v
		}
		if f, err := strconv.ParseFloat(env, 64); err == nil && f > 0 {
			return int64(f)
		}
	}
	return 15 * 60 * 1000
}

// IsFresh is a pure freshness predicate.
func IsFresh(agg *Aggregate, ttl, now int64) bool {
	return agg != nil && (now-agg.GeneratedAt) < ttl
}

func isProjectContextFresh(e *projEntry, ttl, now int64) bool {
	return e != nil && (now-e.generatedAt) < ttl
}

// ReadCacheFile loads a cache file, returning nil if missing/invalid.
func ReadCacheFile(path string) *Aggregate {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var agg Aggregate
	if json.Unmarshal(b, &agg) != nil {
		return nil
	}
	if agg.Projects == nil {
		return nil // mirrors the Array.isArray(projects) guard
	}
	return &agg
}

// WriteCacheFile persists the aggregate (best-effort).
func WriteCacheFile(agg *Aggregate, path string) {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return
	}
	b, err := json.Marshal(agg)
	if err != nil {
		return
	}
	_ = os.WriteFile(path, b, 0o644)
}

// ComputeAggregate collects the global layer + sidebar fresh.
func ComputeAggregate(activeRepo, home string, now int64) Aggregate {
	return Aggregate{
		GeneratedAt: now,
		Global:      collectors.CollectGlobal(home),
		Projects:    collectors.CollectSidebar(activeRepo),
	}
}

var (
	mu          sync.Mutex
	current     *Aggregate
	refreshing  bool
	refreshDone chan struct{}
	started     bool
)

// GetAggregate returns the latest in-memory aggregate (or nil).
func GetAggregate() *Aggregate {
	mu.Lock()
	defer mu.Unlock()
	return current
}

// IsRefreshing reports whether an aggregate refresh is in flight.
func IsRefreshing() bool {
	mu.Lock()
	defer mu.Unlock()
	return refreshing
}

// RefreshAggregate runs a single-flight aggregate computation, blocking until it
// completes (concurrent callers wait on the same run). Call in a goroutine for
// fire-and-forget behavior.
func RefreshAggregate(activeRepo, home string) {
	mu.Lock()
	if refreshing {
		ch := refreshDone
		mu.Unlock()
		<-ch
		return
	}
	refreshing = true
	refreshDone = make(chan struct{})
	ch := refreshDone
	mu.Unlock()

	next := ComputeAggregate(activeRepo, home, nowMs())

	mu.Lock()
	current = &next
	refreshing = false
	refreshDone = nil
	mu.Unlock()

	WriteCacheFile(&next, CacheFile())
	close(ch)
}

// EnsureFresh kicks off a refresh ONLY if the cache is stale AND none is running.
// Non-blocking, fire-and-forget.
func EnsureFresh(activeRepo, home string) {
	mu.Lock()
	if !started || refreshing || IsFresh(current, TTLMs(), nowMs()) {
		mu.Unlock()
		return
	}
	mu.Unlock()
	go RefreshAggregate(activeRepo, home)
}

// StartCache primes once on boot: load a fresh cache file if present, else
// compute one aggregate (single-flight). No background timer.
func StartCache(activeRepo, home string) {
	mu.Lock()
	started = true
	mu.Unlock()
	fromFile := ReadCacheFile(CacheFile())
	if IsFresh(fromFile, TTLMs(), nowMs()) {
		mu.Lock()
		current = fromFile
		mu.Unlock()
		return
	}
	RefreshAggregate(activeRepo, home)
}

// ---------------------------------------------------------------------------
// Per-project detail cache
// ---------------------------------------------------------------------------

type projEntry struct {
	generatedAt int64
	context     map[string]any
}

type projCall struct {
	done   chan struct{}
	result map[string]any
}

var (
	pcMu                     sync.Mutex
	projectContextCache      = map[string]projEntry{}
	projectContextInFlight   = map[string]*projCall{}
	projectContextGeneration = map[string]int{}
)

// GetProjectContext returns a cached project context or starts a single-flight
// computation. projectKey is the resolved absolute project path.
func GetProjectContext(projectKey string, computeFn func() map[string]any) map[string]any {
	mu.Lock()
	st := started
	mu.Unlock()
	if !st {
		return computeFn() // test mode: always live
	}

	now := nowMs()
	ttl := TTLMs()
	pcMu.Lock()
	if e, ok := projectContextCache[projectKey]; ok && isProjectContextFresh(&e, ttl, now) {
		pcMu.Unlock()
		return e.context
	}
	if c, ok := projectContextInFlight[projectKey]; ok {
		pcMu.Unlock()
		<-c.done
		return c.result
	}
	genAtStart := projectContextGeneration[projectKey]
	c := &projCall{done: make(chan struct{})}
	projectContextInFlight[projectKey] = c
	pcMu.Unlock()

	ctx := computeFn()

	pcMu.Lock()
	if projectContextGeneration[projectKey] == genAtStart {
		projectContextCache[projectKey] = projEntry{generatedAt: nowMs(), context: ctx}
	}
	delete(projectContextInFlight, projectKey)
	c.result = ctx
	close(c.done)
	pcMu.Unlock()
	return ctx
}

// InvalidateProjectContext drops a project's cache entry + bumps its generation
// (so a late-finishing compute discards its stale result).
func InvalidateProjectContext(projectKey string) {
	pcMu.Lock()
	delete(projectContextCache, projectKey)
	projectContextGeneration[projectKey]++
	pcMu.Unlock()
}

// PrimeProjectContext computes the start project's context once on boot.
func PrimeProjectContext(cwd, claudeHome string, agg *Aggregate) {
	mu.Lock()
	st := started
	mu.Unlock()
	if !st {
		return
	}
	pcMu.Lock()
	e, ok := projectContextCache[cwd]
	fresh := ok && isProjectContextFresh(&e, TTLMs(), nowMs())
	pcMu.Unlock()
	if fresh {
		return
	}
	var projects []collectors.SidebarProject
	var global map[string]any
	if agg != nil {
		projects = agg.Projects
		global = agg.Global
	}
	GetProjectContext(cwd, func() map[string]any {
		return appctx.BuildContext(cwd, claudeHome, appctx.BuildOpts{
			Projects:      projects,
			ActiveProject: lastSeg(cwd),
			Global:        global,
			SkipProject:   false,
		})
	})
}

func lastSeg(p string) string {
	for i := len(p) - 1; i >= 0; i-- {
		if p[i] == '/' {
			return p[i+1:]
		}
	}
	return p
}

// Test helpers.
func ResetProjectContextCacheForTest() {
	pcMu.Lock()
	projectContextCache = map[string]projEntry{}
	projectContextInFlight = map[string]*projCall{}
	projectContextGeneration = map[string]int{}
	pcMu.Unlock()
}

func SetStartedForTest(v bool) {
	mu.Lock()
	started = v
	mu.Unlock()
}
